syntax = "proto3";

option java_package = "raft.grpc";

package raftcpp;

message RequestVoteRequest {
    int32 term = 1;
    string candidateId = 2;
    int32 lastLogIndex = 3;
    int32 lastLogTerm = 4;
}

message RequestVoteReply {
    int32 term = 1;
    bool voteGranted = 2;
}

message AppendEntriesRequest {
    int32 term = 1;
    string leaderId = 2;
    int32 prevLogIndex = 3;
    int32 prevLogTerm = 4;
    repeated string entries = 5;
    int32 leaderCommit = 6;
}

message AppendEntriesReply {
    int32 term = 1;
    bool success = 2;
}

message IstallSnapshotRequest {
    int32 term =  1;
    int32 leaderId = 2;
    int32 lastIncludedIndex = 3;
    int32 lastIncludedTerm = 4;
    int32 offset = 5;
    repeated string data = 6;
    bool done = 7;
}

message InstallSnapshotReply {
    int32 term = 1;
}

service RaftService {
    rpc RequestVote(RequestVoteRequest) returns (RequestVoteReply) {}
    rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesReply) {} 
    rpc InstallSnapshot(IstallSnapshotRequest) returns  (InstallSnapshotReply) {}
}

